lib/Marlin/Marlin/src/module/motion.cpp

```diff
@@ -1166,44 +1166,31 @@ void prepare_move_to_destination() {
 
 #if HAS_ENDSTOPS
 
-  /**
-   * axis_homed
-   *   Flags that each linear axis was homed.
-   *   XYZ on cartesian, ABC on delta, ABZ on SCARA.
-   *
-   * axis_known_position
-   *   Flags that the position is known in each linear axis. Set when homed.
-   *   Cleared whenever a stepper powers off, potentially losing its position.
-   */
-  uint8_t axis_homed, axis_known_position; // = 0
-
-  uint8_t axes_need_homing(uint8_t axis_bits/*=0x07*/) {
-    #if ENABLED(HOME_AFTER_DEACTIVATE)
-      #define HOMED_FLAGS axis_known_position
-    #else
-      #define HOMED_FLAGS axis_homed
-    #endif
-    // Clear test bits that are homed
-    if (TEST(axis_bits, X_AXIS) && TEST(HOMED_FLAGS, X_AXIS)) CBI(axis_bits, X_AXIS);
-    if (TEST(axis_bits, Y_AXIS) && TEST(HOMED_FLAGS, Y_AXIS)) CBI(axis_bits, Y_AXIS);
-    if (TEST(axis_bits, Z_AXIS) && TEST(HOMED_FLAGS, Z_AXIS)) CBI(axis_bits, Z_AXIS);
+  main_axes_bits_t axes_homed, axes_trusted; // = 0
+
+  main_axes_bits_t axes_should_home(main_axes_bits_t axis_bits/*=main_axes_mask*/) {
+    auto set_should = [](main_axes_bits_t &b, AxisEnum a) {
+      if (TEST(b, a) && TERN(HOME_AFTER_DEACTIVATE, axis_is_trusted, axis_was_homed)(a))
+        CBI(b, a);
+    };
+    // Clear test bits that are trusted
+    NUM_AXIS_CODE(
+      set_should(axis_bits, X_AXIS), set_should(axis_bits, Y_AXIS), set_should(axis_bits, Z_AXIS),
+      set_should(axis_bits, I_AXIS), set_should(axis_bits, J_AXIS), set_should(axis_bits, K_AXIS),
+      set_should(axis_bits, U_AXIS), set_should(axis_bits, V_AXIS), set_should(axis_bits, W_AXIS)
+    );
     return axis_bits;
   }
 
-  bool axis_unhomed_error(uint8_t axis_bits/*=0x07*/) {
-    if ((axis_bits = axes_need_homing(axis_bits))) {
+  bool homing_needed_error(main_axes_bits_t axis_bits/*=main_axes_mask*/) {
+    if ((axis_bits = axes_should_home(axis_bits))) {
       PGM_P home_first = GET_TEXT(MSG_HOME_FIRST);
-      char msg[strlen_P(home_first)+1];
-      sprintf_P(msg, home_first,
-        TEST(axis_bits, X_AXIS) ? "X" : "",
-        TEST(axis_bits, Y_AXIS) ? "Y" : "",
-        TEST(axis_bits, Z_AXIS) ? "Z" : ""
-      );
+      char msg[30];
+      #define _AXIS_CHAR(N) TEST(axis_bits, _AXIS(N)) ? STR_##N : ""
+      sprintf_P(msg, home_first, MAPLIST(_AXIS_CHAR, MAIN_AXIS_NAMES));
       SERIAL_ECHO_START();
       SERIAL_ECHOLN(msg);
-      #if HAS_DISPLAY
-        ui.set_status(msg);
-      #endif
+      ui.set_status(msg);
```

Comment:
Isn't there any side effect with this change? Is ui.set_status() always available?

Reply:
In marlin 2.1.x the display logic is more decoupled from the LCD. 2.1.x now always calls the "ui" abstraction, which then dispatches the call to the real display. As such, the UI calls are generally always done, so that code that is completely headless (say, remote display) can still catch the UI and redirect it as needed.

For us there's no functional change, since HAS_DISPLAY was always already set.